app/upgrade_handler.go

```diff
@@ -241,3 +255,120 @@ func transAddrPrefixForGov(ctx sdk.Context, app ShentuApp) (err error) {
 	})
 	return err
 }
+
+func runSlashingMigration(ctx sdk.Context, app ShentuApp) (err error) {
+	sk := app.SlashingKeeper
+	sk.IterateValidatorSigningInfos(ctx, func(address sdk.ConsAddress, info slashingtypes.ValidatorSigningInfo) (stop bool) {
+		info.Address, err = common.PrefixToShentu(info.Address)
+		if err != nil {
+			return true
+		}
+		sk.SetValidatorSigningInfo(ctx, address, info)
+
+		return false
+	})
+
+	return err
+}
+
+func runAuthMigration(ctx sdk.Context, app ShentuApp) (err error) {
+	ak := app.AccountKeeper
+	ak.IterateAccounts(ctx, func(acc sdkauthtypes.AccountI) (stop bool) {
+		switch account := acc.(type) {
+		case *sdkauthtypes.BaseAccount:
+			var newAddr string
+			newAddr, err = common.PrefixToShentu(account.Address)
+			if err != nil {
+				return true
+			}
+			account.Address = newAddr
+			ak.SetAccount(ctx, account)
+		case *sdkauthtypes.ModuleAccount:
+			var newAddr string
+			newAddr, err = common.PrefixToShentu(account.Address)
+			if err != nil {
+				return true
+			}
+			account.Address = newAddr
+			ak.SetAccount(ctx, account)
+		case *authtypes.ManualVestingAccount:
+			var newAddr string
+			newAddr, err = common.PrefixToShentu(account.Address)
+			if err != nil {
+				return true
+			}
+			var newUnlocker string
+			newUnlocker, err = common.PrefixToShentu(account.Unlocker)
+			if err != nil {
+				return true
+			}
+			account.Address = newAddr
+			account.Unlocker = newUnlocker
+			ak.SetAccount(ctx, account)
+		default:
+			err = errors.New("unknown account type")
+			return true
+		}
+		return false
+	})
+	return err
+}
+
+func runAuthzMigration(ctx sdk.Context, app ShentuApp) (err error) {
+	ak := app.AuthzKeeper
+	ak.IterateGrants(ctx, func(granterAddr sdk.AccAddress, granteeAddr sdk.AccAddress, grant authz.Grant) bool {
+		authorization := grant.Authorization
+
+		switch authorization.GetCachedValue().(type) {
+		case *authz.GenericAuthorization:
+		case *stakingtypes.StakeAuthorization:
+			stakeAuthorization := &stakingtypes.StakeAuthorization{}
+			if err = stakeAuthorization.Unmarshal(authorization.Value); err != nil {
+				return true
+			}
+			if err = processStakeAuthorization(stakeAuthorization); err != nil {
+				return true
+			}
+			if err := ak.SaveGrant(ctx, granteeAddr, granterAddr, stakeAuthorization, grant.Expiration); err != nil {
+				return true
+			}
+		default:
```

Comment:
just for your info:
it seems there is another `Authorization`, which is `bank.types.SendAuthorization`. But it looks like we don't have it on mainnet state.

Reply:
Yes, the current data on the chain does not have this status.

Reply:
Add